### PR TITLE
Hybrid-Overlay: Implementing support for VXLAN UDP port configuration for Windows nodes

### PR DESCRIPTION
**Pull request by JocelynBerrendonner**
Signed-off-by: JocelynBerrendonner <jocelynb@microsoft.com>

**- What this PR does and why is it needed**
In order to avoid a UDP port collision on setups where both Hybrid-Overlay and vSphere are installed, the UDP port used by Hybrid-Overlay for VXLAN must be made configurable. This PR captures the changes on the Windows side and exposes a new command line flag to Hybrid-Overlay-Node. This change was made as per RedHat request to support VXLAN UDP port configuration on Windows nodes.

**- Special notes for reviewers**
The Windows implementation of the code, as well as the go-controller config were modified in this change. Additionally, Hybrid-Overlay was re-vendored to pick up a necessary change in Microsoft's Hcsshim API.
It is important to note that this change only covers the Windows side of the implementation. The Linux side needs to be implemented for the end-to-end solution to work seamlessly.

**- How to verify it**
This change was manually verified by running it in an OpenShift cluster with two Linux and two Windows workload nodes. Simple nginx pods where deployed on Linux nodes, and IIS webserver PODs were deployed on Windows nodes. Each Windows pod was able to curl to other pods. Linux nodes were not fully functional (as noted above, the implementation is currently not covering the required Linux changes). TCP packet tracing was enabled on Windows nodes and TCP dump was started on Linux nodes.

**- Description for the changelog**
Exposing the vxlan-port parameter in Hybrid-Overlay and adding the Windows side of the implementation.

### CODE REVIEW

**Comment by JocelynBerrendonner** [comment]
@dcbw : heads up

**Comment by JocelynBerrendonner** [comment]
@JacobTanenbaum : heads up

File: go-controller/hybrid-overlay/pkg/controller/overlay_windows.go [hunk 1]
```diff
@@ -98,6 +99,30 @@ func (info *NetworkInfo) GetHostComputeNetworkConfig() (*hcn.HostComputeNetwork,
...
 
+	policies := []hcn.NetworkPolicy {{
+		Type:     hcn.AutomaticDNS,
+		Settings: dnsJSON,
+	}}
+	
+	// Only configure the VXLAN UDP port if the port is not
+	// the default port.
+	// Note: the calling code is responsible for making sure non-default
+	// ports are only passed to this function if the unrelying platform
```
Comment (mkostersitz) [inline go-controller/hybrid-overlay/pkg/controller/overlay_windows.go:110]: nitpick: underlying

File: go-controller/pkg/config/config.go [hunk 2]
```diff
@@ -95,6 +95,7 @@ var (
 	// HybridOverlay holds hybrid overlay feature config options.
 	HybridOverlay = HybridOverlayConfig{
 		RawClusterSubnets: "10.132.0.0/14/23",
+		VXLANPort: 4789, // 4789 is the default IANA-assigned VXLAN UDP port number
```
Comment (mkostersitz) [inline go-controller/pkg/config/config.go:98]: Do we need to hardcode this twice? Could node.go not pick that up from config.go?

File: go-controller/hybrid-overlay/pkg/controller/node_windows.go [hunk 3]
```diff
@@ -20,6 +21,7 @@ import (
 const (
 	// Hard-coded constants
 	networkName = "OVNKubernetesHybridOverlayNetwork" // In practice, this is the virtual switch name
+	defaultVxlanPort = 4786 // IANA assigned UDP port for VXLAN
```
Comment (JocelynBerrendonner) [inline go-controller/hybrid-overlay/pkg/controller/node_windows.go:24]: There is a typo here: I Need to fix the port

File: go-controller/pkg/config/config.go (hunk 2 above)
Comment (JocelynBerrendonner) [reply go-controller/pkg/config/config.go:98]: Generally speaking, yes, I prefer have one hard-coded location for these constants. The pattern used in config.go seem to indicate the constants were coded all over, so I tried to follow the existing pattern. That said, I found a precedence in using values from config.go from other places, so your suggestion makes sense. I will fix.

**Comment by dcbw** [comment]
@JocelynBerrendonner 

```
!!! 'gofmt -s' needs to be run on the following files: 
./hybrid-overlay/pkg/controller/overlay_windows.go
./hybrid-overlay/pkg/controller/node_windows.go
./pkg/config/config.go
```

File: go-controller/hybrid-overlay/pkg/controller/node_windows.go [hunk 4]
```diff
@@ -100,6 +107,10 @@ func ensureBaseNetwork() error {
 		return nil
 	}
 
+	if (config.HybridOverlay.VXLANPort > 65535) {
```
Comment (dcbw) [inline go-controller/hybrid-overlay/pkg/controller/node_windows.go:109]: @JocelynBerrendonner don't need () around the comparison here

File: go-controller/hybrid-overlay/pkg/controller/node_windows.go (hunk 3 above)
Comment (dcbw) [inline go-controller/hybrid-overlay/pkg/controller/node_windows.go]: Let's do `defaultVXLANPort`

File: go-controller/pkg/config/config.go [hunk 5]
```diff
@@ -762,6 +765,12 @@ var HybridOverlayFlags = []cli.Flag{
 			"hostsubnetlength defines how many IP addresses are dedicated to each node.",
 		Destination: &cliConfig.HybridOverlay.RawClusterSubnets,
 	},
+	&cli.UintFlag{
+		Name:        "vxlan-port",
```
Comment (dcbw) [inline go-controller/pkg/config/config.go:772]: `hybrid-overlay-vxlan-port`

File: go-controller/hybrid-overlay/pkg/controller/node_windows.go [hunk 6]
```diff
@@ -237,6 +254,10 @@ func (n *NodeController) initSelf(node *kapi.Node, nodeSubnet *net.IPNet) error
 	// as to what this gateway address should be.
 	gatewayAddress := util.NextIP(nodeSubnet.IP)
 
+	if (config.HybridOverlay.VXLANPort > 65535) {
```
Comment (dcbw) [inline go-controller/hybrid-overlay/pkg/controller/node_windows.go:256]: Don't  need the () around the condition here.

**Comment by dcbw** [comment]
mostly LGTM except for a few nits

**Comment by JocelynBerrendonner** [comment]
Thanks @dcbw ! I addressed your feedback, ran gofmt, and merged the commits together.

**Pull request merged**